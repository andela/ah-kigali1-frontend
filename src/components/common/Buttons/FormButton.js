import React from "react";
import PropTypes from "prop-types";
const FormButton = props => {
  return (
    <input
      {...props}
      type="button"
      value={props.value}
      onClick={props.onClick}
      disabled={props.disabled}
    />
  );
};

FormButton.prototype = {
  value: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool
};
FormButton.defaultProps = {
  value: "Submit",
  disabled: false
};
export default FormButton;
