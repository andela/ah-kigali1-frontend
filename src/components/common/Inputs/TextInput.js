import React from "react";
import PropTypes from "prop-types";

const TextInput = ({
  name,
  type,
  id,
  placeholder,
  value,
  onChange,
  ...props
}) => (
  <input
    {...props}
    name={name}
    type={type}
    id={id}
    placeholder={placeholder}
    value={value}
    onChange={onChange}
  />
);

TextInput.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  id: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: "",
  onKeyDown: () => ""
};

export default TextInput;
