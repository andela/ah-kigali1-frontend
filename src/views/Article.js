import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import addArticle from "../redux/actions/index";
// import "../index.css";

const mapDispatchToProps = dispatch => ({
  addArticle: () => dispatch(addArticle())
});

const FetchedArticles = props => (
  <div>
    <h1>Fetch some articles from dummy app</h1>

    <button
      // eslint-disable-next-line react/destructuring-assignment
      onClick={() => props.addArticle()}
      className="btn"
      type="submit"
    >
      Fetch more articles
    </button>
  </div>
);

FetchedArticles.propTypes = {
  // eslint-disable-next-line react/require-default-props
  addArticle: PropTypes.func
};

const Articles = connect(
  null,
  mapDispatchToProps
)(FetchedArticles);

export default Articles;
