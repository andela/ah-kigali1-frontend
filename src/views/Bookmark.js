import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import BookmarkComponent from "../components/common/Bookmarks";
import bookmarkIcon from "../assets/img/bookmark-icons.svg";
import bookmarkedIcon from "../assets/img/bookmarked.svg";
import { handleBookmark } from "../redux/actions/bookmarkActions";

export class Bookmark extends Component {
  static chooseBookmarkIcon(condition) {
    return condition ? bookmarkedIcon : bookmarkIcon;
  }

  render() {
    const {
      bookmark,
      isBookmarked,
      slug,
      isSubmitting,
      bookmarkedArticles,
      location,
      history
    } = this.props;
    if (!Object.prototype.hasOwnProperty.call(bookmarkedArticles, slug)) {
      Object.assign(bookmarkedArticles, { [slug]: isBookmarked });
    }
    const theBookmarkIcon = Bookmark.chooseBookmarkIcon(
      bookmarkedArticles[slug]
    );
    return (
      <BookmarkComponent
        onClick={() =>
          bookmark(slug, bookmarkedArticles[slug], { location, history })
        }
        icon={theBookmarkIcon}
        disabled={isSubmitting}
      />
    );
  }
}

export const mapStateToProps = state => ({
  ...state.bookmark
});

export const mapDispatchToProps = dispatch => ({
  bookmark: (slug, isBookmarked, { history, location }) =>
    dispatch(handleBookmark(slug, isBookmarked, { history, location }))
});

Bookmark.propTypes = {
  bookmark: PropTypes.func.isRequired,
  isBookmarked: PropTypes.bool.isRequired,
  bookmarkedArticles: PropTypes.shape({}).isRequired,
  slug: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  location: PropTypes.shape([]).isRequired,
  history: PropTypes.shape({}).isRequired
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Bookmark)
);
