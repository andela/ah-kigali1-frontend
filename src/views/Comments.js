/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable no-shadow */
import React, { Component } from "react";
import Textarea from "react-textarea-autosize";
import PropTypes from "prop-types";
import moment from "moment";
import authorImage from "../assets/img/user.jpg";
import Button from "../components/common/Buttons/BasicButton";
import DeleteDialogue from "../components/common/Dialog/DeleteDialogue";
import heartIcon from "../assets/img/heart.svg";

export class Comments extends Component {
  state = {
    editMode: false,
    index: null
  };

  onEnterPress = (e, id) => {
    const { editComment } = this.props;
    if (e.keyCode === 13 && e.shiftKey === false) {
      e.preventDefault();
      editComment(id);
      this.setState({ editMode: false });
    }
  };

  isEditing = (body, index) => {
    const { onSetBodyEdit } = this.props;
    this.setState({ editMode: true, index });
    onSetBodyEdit({ value: body });
  };

  handleComment = id => {
    const { editComment } = this.props;
    editComment(id);
    this.setState({ editMode: false });
  };

  render() {
    const { editMode, index: commentIndex } = this.state;
    const {
      comments,
      updatedBody,
      deleteComment,
      handleCommentsInputEdit
    } = this.props;

    return Object.values(comments).length
      ? Object.values(comments).map((comment, index) => {
          const { body, like, id, createdAt, author } = comment;
          const { image, lastName, firstName, username } = author;
          return (
            <div className="article-comments--existing__desktop" key={id}>
              <div className="avatar-wrapper comment-avatar-wrapper">
                <img
                  src={image || authorImage}
                  alt="Avatar"
                  className="avatar"
                />
                <span className="comment-author_name ">
                  {username && firstName && lastName
                    ? `${firstName} ${lastName.charAt(0) ||
                        username.charAt(0)}.`
                    : username}
                </span>
              </div>
              <div className="article-comments--existing-text">
                {editMode && Number(commentIndex) === index ? (
                  <div>
                    <Textarea
                      id="edit-comment-textarea"
                      type="textarea"
                      onChange={handleCommentsInputEdit}
                      onKeyDown={e => this.onEnterPress(e, id)}
                      name="bodyEdit"
                      value={updatedBody}
                    />
                    <Button
                      id="update-comment"
                      className="btn delete_article"
                      onClick={() => this.handleComment(id)}
                      title="Update"
                    />
                    <Button
                      id="close-update-comment"
                      className="btn delete_article"
                      onClick={() => this.setState({ editMode: false })}
                      title="Cancel"
                    />
                  </div>
                ) : (
                  <div>
                    <div className="time-ago">
                      Posted {moment(createdAt).fromNow()}
                    </div>
                    <div className="comment-body" key={id}>
                      {body}
                    </div>
                    <div className="comment-edit-delete">
                      <DeleteDialogue
                        id="delete-comment"
                        deleteComment={() => deleteComment(id)}
                      />
                      <div
                        className="edit-comment"
                        onClick={() => this.isEditing(body, index)}
                      >
                        <i className="fa fa-edit" />
                      </div>
                    </div>
                    <div className="article-comments--actions">
                      <span className="comment-like">
                        <div className="icons">
                          <img src={heartIcon} alt="likes" className="likes" />
                          <div>{like}</div>
                        </div>
                      </span>
                    </div>
                  </div>
                )}
              </div>
            </div>
          );
        })
      : "";
  }
}

Comments.propTypes = {
  comments: PropTypes.shape([]).isRequired,
  updatedBody: PropTypes.string.isRequired,
  onSetBodyEdit: PropTypes.func.isRequired,
  deleteComment: PropTypes.func.isRequired,
  editComment: PropTypes.func.isRequired,
  handleCommentsInputEdit: PropTypes.func.isRequired
};

export default Comments;
