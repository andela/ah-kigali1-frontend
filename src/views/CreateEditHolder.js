import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import RichTextEditor from "react-rte";
import Validator from "../utils/validator";
import { checkNextProps } from "../utils/helperFunctions";
import Input from "../components/common/Inputs/TextInput";
import Button from "../components/common/Buttons/BasicButton";
import Error from "../components/common/Message/error";
import Tag from "../components/common/Tags/NewTag";
import { toolbarConfig } from "../utils/editor/editorConfig";
import {
  editArticle,
  newArticle,
  fetchOneArticle
} from "../redux/actions/newArticle";

export const mapDispatchToProps = dispatch => ({
  fetchArticleToEdit: slug => dispatch(fetchOneArticle(slug)),
  createArticle: (article, history) => dispatch(newArticle(article, history)),
  editOneArticle: (article, slug, history) =>
    dispatch(editArticle(article, slug, history))
});

export const mapStateToProps = state => ({
  ...state.newArticle,
  ...state.auth
});

export class Article extends Component {
  constructor(props) {
    super(props);
    this.state = {
      edit: false,
      tag: "",
      title: "",
      isFetching: false,
      description: "",
      tagsList: [],
      error: "",
      value: RichTextEditor.createEmptyValue()
    };
  }

  componentDidMount = () => {
    const {
      match: {
        params: { slug }
      },
      fetchArticleToEdit,
      history: {
        location: { pathname }
      }
    } = this.props;
    if (pathname !== "/articles/new") {
      this.setState({ edit: true, isFetching: true });
      fetchArticleToEdit(slug);
    }
  };

  componentWillReceiveProps = nextProps => {
    const { articleError } = nextProps;
    if (articleError) {
      this.setState({ error: articleError });
      return setTimeout(() => this.setState({ error: "" }), 3000);
    }
    const { article } = this.props;

    if (checkNextProps(nextProps.article, article)) {
      this.setState({
        ...nextProps.article,
        isFetching: false,
        value: RichTextEditor.createValueFromString(
          nextProps.article.body,
          "html"
        )
      });
    }
  };

  onEditorChange = value => {
    this.setState({ value });
  };

  onRemoveTag = value => {
    const { tagsList } = this.state;
    const index = tagsList.indexOf(value);
    tagsList.splice(index, 1);
    this.setState({ tagsList });
  };

  cancelEdit = () => {
    const { history } = this.props;
    return history.push("/");
  };

  handleInputChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  submitArticle = () => {
    const {
      editOneArticle,
      createArticle,
      history,
      history: {
        location: { pathname }
      },
      match: {
        params: { slug }
      }
    } = this.props;
    const { title, value, description } = this.state;

    const error = Validator.newArticleValidation({
      title,
      body: value.toString("html"),
      description
    });
    if (error) {
      this.setState({ error });
      return this.clearErrorMessage();
    }
    this.setState({
      error: ""
    });
    if (pathname !== "/articles/new") {
      return editOneArticle(
        {
          title,
          body: value.toString("html"),
          description
        },
        slug,
        history
      );
    }
    return createArticle(
      {
        title,
        body: value.toString("html"),
        description
      },
      history
    );
  };

  handleNewTag = event => {
    const { tag, tagsList } = this.state;
    if (event.keyCode === 13 && tag.length) {
      this.setState({ tag: "", tagsList: [...tagsList, tag] });
    }
  };

  clearErrorMessage = () =>
    setTimeout(() => this.setState({ error: "" }), 3000);

  displayErrorMessage = () => {
    const { error } = this.state;
    return (
      <div className="error error__message">
        {error.length ? <Error title={error} /> : ""}
      </div>
    );
  };

  render() {
    const {
      value,
      edit,
      tag,
      title,
      body,
      description,
      tagsList,
      isFetching
    } = this.state;

    const { isSubmitting } = this.props;
    return isFetching ? (
      ""
    ) : (
      <div>
        {this.displayErrorMessage()}
        <div className="new-article-container container">
          <div className="buttons">
            {edit ? (
              <Button
                className="button--dark cancel"
                title="Cancel"
                disabled={isSubmitting}
                onClick={this.cancelEdit}
              />
            ) : (
              ""
            )}
            <Button
              className="button--dark publish"
              title={edit ? "Save" : "Publish"}
              disabled={isSubmitting}
              onClick={this.submitArticle}
            />
          </div>
          <div className="main-article">
            <div className="title">
              <Input
                className="title__input"
                type="text"
                value={title}
                placeholder="The title should come here"
                name="title"
                onChange={e => this.handleInputChange(e)}
              />
            </div>
            <hr className="line__title" />
            <div className="description">
              <Input
                className="description__input"
                type="text"
                value={description}
                placeholder="Description should be here"
                name="description"
                onChange={e => this.handleInputChange(e)}
              />
            </div>
            <hr className="line__title" />
            <div className="tags">
              {tagsList.map(newTag => (
                <Tag
                  title={newTag}
                  key={newTag}
                  onClick={() => this.onRemoveTag(newTag)}
                />
              ))}
              <Input
                type="text"
                id="new-tag"
                placeholder="new tag"
                name="tag"
                className="tag"
                value={tag}
                onKeyDown={e => this.handleNewTag(e)}
                onChange={e => this.handleInputChange(e)}
              />
            </div>
            <RichTextEditor
              editorClassName="demo_editor"
              placeholder={body}
              value={value}
              onChange={e => this.onEditorChange(e)}
              toolbarConfig={toolbarConfig}
            />
          </div>
        </div>
      </div>
    );
  }
}

Article.propTypes = {
  editOneArticle: PropTypes.func.isRequired,
  createArticle: PropTypes.func.isRequired,
  fetchArticleToEdit: PropTypes.func.isRequired,
  articleError: PropTypes.shape({}).isRequired,
  isSubmitting: PropTypes.bool,
  article: PropTypes.shape({
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired,
    message: PropTypes.string,
    response: PropTypes.object,
    description: PropTypes.string,
    tagsList: PropTypes.arrayOf(PropTypes.string)
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      slug: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

Article.defaultProps = {
  isSubmitting: false
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Article);
