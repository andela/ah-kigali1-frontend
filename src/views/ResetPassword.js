import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {
  handleInputChange,
  sendResetLink
} from "../redux/actions/resetPasswordActions";
import TextInput from "../components/common/Inputs/TextInput";
import FormButton from "../components/common/Buttons/FormButton";

export class ResetPassword extends Component {
  constructor(props) {
    super(props);
    this.handleOnChange = this.handleOnChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleOnChange(e) {
    const { name, value } = e.target;
    const { handleInputChange: handleChange } = this.props;
    handleChange(name, value);
  }

  handleSubmit() {
    const { email } = this.props;
    const { sendResetLink: handleReset } = this.props;
    handleReset({ email });
  }

  render() {
    const {
      email,
      failedMessage,
      isSubmitting,
      isSuccess,
      successMessage
    } = this.props;
    return (
      <div className="reset-password__container">
        <div className="reset-password">
          <div className="reset-password__header">
            <img
              src={require("../assets/img/main-logo.svg")}
              alt="logo"
              className="logo"
            />
            <h1>Author &apos;s Heaven</h1>
            {isSuccess ? (
              <div>
                <h2>Email sent!</h2>
                <p>{successMessage}</p>
              </div>
            ) : (
              <div>
                <h2>Reset password</h2>
                <p>
                  To reset your password, please provide your Authorâ€™s Heaven
                  email.
                </p>
                <form action="#" className="reset-password__form">
                  <TextInput
                    type="email"
                    name="email"
                    placeholder="Email"
                    onChange={e => this.handleOnChange(e)}
                    value={email}
                  />
                  {failedMessage ? (
                    <div className="auth-errors">
                      <p className="danger">{failedMessage}</p>
                    </div>
                  ) : (
                    ""
                  )}
                  <FormButton
                    value="Sign In"
                    disabled={isSubmitting}
                    onClick={() => this.handleSubmit()}
                  />
                </form>
              </div>
            )}
          </div>
          <div className="reset-password__footer">
            <p>
              Don &apos;t have an account?
              <a href="/">sign up now!</a>
            </p>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => {
  const { resetPassword } = state;
  return {
    ...resetPassword
  };
};
ResetPassword.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  email: PropTypes.string.isRequired,
  sendResetLink: PropTypes.func.isRequired,
  failedMessage: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  isSuccess: PropTypes.bool.isRequired,
  successMessage: PropTypes.string.isRequired
};
export default connect(
  mapStateToProps,
  { handleInputChange, sendResetLink }
)(ResetPassword);
