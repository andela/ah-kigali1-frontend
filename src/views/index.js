import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import Login from "./Login";
import ResetPassword from "./ResetPassword";
import Home from "./Home";
import UpdatePassword from "./UpdatePassword";
import SocialAuth from "./SocialAuth";
import NewArticle from "./NewArticle";
import EditArticle from "./EditArticle";
import NotFound from "./NotFound";
import ReadArticle from "./ReadArticle";

const Routers = () => (
  <Router>
    <Switch>
      <Route exact path="/" component={Home} />
      <Route exact path="/sign_in" component={Login} />
      <Route exact path="/social_auth" component={SocialAuth} />
      <Route exact path="/reset_password" component={ResetPassword} />
      <Route exact path="/update_password" component={UpdatePassword} />
      <Route exact path="/articles/:slug/edit" component={EditArticle} />
      <Route exact path="/articles/new_article" component={NewArticle} />
      <Route exact path="/not_found" component={NotFound} />
      <Route exact path="/articles/:slug" component={ReadArticle} />
    </Switch>
  </Router>
);
export default Routers;
