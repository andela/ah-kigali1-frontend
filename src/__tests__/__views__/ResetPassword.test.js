import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import { ResetPassword } from "../../views/ResetPassword";
import TextInput from "../../components/common/Inputs/TextInput";
import FormButton from "../../components/common/Buttons/FormButton";

const [handleInputChange, sendResetLink] = new Array(2).fill(jest.fn());
const userInput = { email: "me@example.com" };
const setUp = () => {
  const props = {
    handleInputChange,
    email: userInput.email,
    sendResetLink,
    failedMessage: "",
    isSubmitting: false,
    isSuccess: false,
    successMessage: ""
  };
  const wrapper = shallow(<ResetPassword {...props} />);
  return { props, wrapper };
};
describe("Login component", () => {
  let component;
  beforeEach(() => {
    sinon.spy(ResetPassword.prototype, "handleOnChange");
    sinon.spy(ResetPassword.prototype, "handleSubmit");
    const { wrapper } = setUp();
    component = wrapper;
  });
  afterEach(() => {
    ResetPassword.prototype.handleOnChange.restore();
    ResetPassword.prototype.handleSubmit.restore();
  });
  describe("component render", () => {
    it("should render two TextInputs and button", () => {
      expect(component.find(TextInput).length).toBe(1);
      expect(component.find(FormButton).length).toBe(1);
    });
    it("should render component with initial props", () => {
      component.find(TextInput).forEach(input => {
        expect(input.props().value).toBe(userInput.email);
      });
    });
    it("should call handleOnChange when user enter input", () => {
      component
        .find(TextInput)
        .at(0)
        .simulate("change", {
          target: {
            name: "email",
            value: "example@yahoo.com"
          }
        });

      expect(ResetPassword.prototype.handleOnChange.calledOnce).toBe(true);
      expect(handleInputChange).toBeCalledWith("email", "example@yahoo.com");
    });

    it("returns validation errors", () => {
      component
        .find(FormButton)
        .at(0)
        .simulate("click");
      expect(ResetPassword.prototype.handleSubmit.calledOnce).toBe(true);
      expect(sendResetLink).toBeCalledWith({ ...userInput });
    });
  });
});
