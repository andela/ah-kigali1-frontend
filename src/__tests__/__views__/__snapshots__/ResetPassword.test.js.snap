// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Password component snapshot returns the snapshot 1`] = `
<div
  className="reset-password__container"
>
  <div
    className="reset-password"
  >
    <div
      className="reset-password__header"
    >
      <img
        alt="logo"
        className="logo"
        src="test-file-stub"
      />
      <h1>
        Author 's Heaven
      </h1>
      <div>
        <h2>
          Reset password
        </h2>
        <p>
          To reset your password, please provide your Authorâ€™s Heaven email.
        </p>
        <form
          action="#"
          className="reset-password__form"
        >
          <TextInput
            id=""
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value="me@example.com"
          />
          <div
            className="auth-errors"
          >
            <p
              className="danger"
            />
          </div>
          <FormButton
            disabled={false}
            onClick={[Function]}
            value="Sign In"
          />
        </form>
      </div>
    </div>
    <div
      className="reset-password__footer"
    >
      <p>
        Don 't have an account?
        <a
          href="/"
        >
          sign up now!
        </a>
      </p>
    </div>
  </div>
</div>
`;
